File: app/src/main/java/org/miaowo/miaowo/root/BaseRecyclerAdapter.java
package org.miaowo.miaowo.root;

import android.support.v7.widget.RecyclerView;
import android.util.SparseIntArray;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;
import java.util.function.UnaryOperator;

abstract public class BaseRecyclerAdapter<T> extends RecyclerView.Adapter<BaseViewHolder> {
    private SparseIntArray mLayoutIds = new SparseIntArray();
    private List<T> mItems = new ArrayList<>();
    private UnaryOperator<Integer> mTypeChooser;

    public BaseRecyclerAdapter(int layoutId) {
        mLayoutIds.put(0, layoutId);
    }

    public BaseRecyclerAdapter(int[] types, int[] layouts, UnaryOperator<Integer> typeChooser) {
        for (int i = 0; i < types.length; i++) mLayoutIds.put(types[i], layouts[i]);
        mTypeChooser = typeChooser;
    }

    public void update(List<T> items) {
        mItems = items;
        BaseActivity.get.runOnUiThreadIgnoreError(this::notifyDataSetChanged);
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new BaseViewHolder(mLayoutIds.get(viewType), parent);
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public T getItem(int position) {
        return mItems.get(position);
    }

    @Override
    public int getItemViewType(int position) {
        if (mTypeChooser == null) return 0;
        return mTypeChooser.apply(position);
    }
}

File: app/src/main/java/org/miaowo/miaowo/base/BaseRecyclerAdapter.kt
package org.miaowo.miaowo.base

import android.support.v7.widget.RecyclerView
import android.util.SparseIntArray
import android.view.ViewGroup

abstract class BaseRecyclerAdapter<T> : RecyclerView.Adapter<BaseViewHolder> {
    private val mLayoutIds = SparseIntArray()
    private var mItems: List<T> = mutableListOf()
    private var mTypeChooser: (position: Int) -> Int = {0}

    constructor(layoutId: Int) {
        mLayoutIds.put(0, layoutId)
    }

    constructor(types: Array<Int>, layouts: Array<Int>, typeChooser: (position: Int) -> Int) {
        for (i in types.indices) mLayoutIds.put(types[i], layouts[i])
        mTypeChooser = typeChooser
    }

    fun update(items: List<T>) {
        mItems = items
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            BaseViewHolder(mLayoutIds.get(viewType), parent)

    override fun getItemCount() = mItems.size

    fun getItem(position: Int) = mItems[position]

    override fun getItemViewType(position: Int) = mTypeChooser.invoke(position)
}
